SRC/cggqrcs.f

```diff
@@ -0,0 +1,724 @@
+*>          the effective rank of G. Generally, it is set to
+*>                   TOL = MAX( M + P, N ) * norm(G) * MACHEPS,
+*>          where norm(G) is the Frobenius norm of G.
+*>          The size of TOL may affect the size of backward error of the
+*>          decomposition.
+*> \endverbatim
+*
+*  Authors:
+*  ========
+*
+*> \author Christoph Conrads (https://christoph-conrads.name)
+*
+*> \date October 2019, May 2020
+*
+*> \ingroup realGEsing
+*
+*> \par Contributors:
+*  ==================
+*>
+*>     Christoph Conrads (https://christoph-conrads.name)
+*>
+*
+*> \par Further Details:
+*  =====================
+*>
+*>  CGGQRCS should be significantly faster than DGGSVD3 for large
+*>  matrices because the matrices A and B are reduced to a pair of
+*>  well-conditioned bidiagonal matrices instead of pairs of upper
+*>  triangular matrices. On the downside, CGGQRCS requires a much larger
+*>  workspace whose dimension must be queried at run-time. CGGQRCS also
+*>  offers no guarantees which of the two possible diagonal matrices
+*>  is used for the matrix factorization.
+*>
+*  =====================================================================
+      RECURSIVE SUBROUTINE CGGQRCS( JOBU1, JOBU2, JOBX, M, N, P, L,
+     $                              SWAPPED,
+     $                              A, LDA, B, LDB,
+     $                              ALPHA, BETA,
+     $                              U1, LDU1, U2, LDU2,
+     $                              WORK, LWORK, RWORK, LRWORK, IWORK,
+     $                              INFO )
+*
+*  -- LAPACK driver routine (version 3.7.0) --
+*  -- LAPACK is a software package provided by Univ. of Tennessee,    --
+*  -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd. --
+*     September 2016
+*
+      IMPLICIT NONE
+*     .. Scalar Arguments ..
+      LOGICAL            SWAPPED
+      CHARACTER          JOBU1, JOBU2, JOBX
+      INTEGER            INFO, LDA, LDB, LDU1, LDU2, L, M, N, P, LWORK,
+     $                   LRWKOPT, LRWORK, LRWORK2BY1
+*     ..
+*     .. Array Arguments ..
+      INTEGER            IWORK( * )
+      REAL               ALPHA( N ), BETA( N ), RWORK( * )
+      COMPLEX            A( LDA, * ), B( LDB, * ),
+     $                   U1( LDU1, * ), U2( LDU2, * ),
+     $                   WORK( * )
+*     ..
+*
+*  =====================================================================
+*
+*     .. Parameters ..
+      COMPLEX            CONE, CZERO
+      PARAMETER          ( CONE = ( 1.0E0, 0.0E0 ),
+     $                   CZERO = ( 0.0E0, 0.0E0 ) )
+*     .. Local Scalars ..
+      LOGICAL            WANTU1, WANTU2, WANTX, LQUERY
+      INTEGER            I, J, K, K1, LMAX, Z, LDG, LDX, LDVT, LWKOPT
+      REAL               BASE, NAN, NORMA, NORMB, NORMG, TOL, ULP, UNFL,
+     $                   THETA, IOTA, W
+      COMPLEX            CNAN
+*     .. Local Arrays ..
+      COMPLEX            G( M + P, N ), VT( N, N )
```

Comment:
This allocates two local arrays. This must be avoided or we would not need the caller's workspace.

Reply:
This comment applies to all xGGQRCS functions.